plugin/core/sessions.py

```diff
@@ -943,12 +943,16 @@ def on_session_buffer_changed_async(self, session_buffer: SessionBufferProtocol)
         """
         pass
 
-    def on_session_end_async(self) -> None:
+    def on_session_end_async(self, exit_code: Optional[int], exception: Optional[Exception]) -> None:
```

Comment:
Why optional `exit_code`? What does `None` symbolize?

Reply:
I've read the docstring now but I don't think I agree with that. When session hasn't crashed then the exit code of 0 should be perfectly valid.

Reply:
The API is called from two places. Here: https://github.com/sublimelsp/LSP/blob/fd1a1ce115ddcebb13c5fcd9961d870ac3dc02b4/plugin/core/sessions.py#L1856 the server is not yet stopped.

It is explained in the docstring:
> If the session hasn't crashed, a shutdown message will be send immediately
        after this method returns. In this case exit_code and exception are None.
        If the session has crashed, the exit_code or the exception is not None.

Reply:
But I would still report `0`. It's an explicit shutdown of the server not caused by a crash so IMO it's OK to specify 0. Is there more specific reason why you think `None` is better?

Reply:
`None` would allow detecting if the server quit on his own or crashed with exit-code 0 for some reason (`exit_code == 0`) or if it was shutdown by the LSP plugin (`exit_code == None`).

Reply:
I just realized that there is a better reason to differentiate: In case of None the session is still usable and requests to the server can be made. In case of not None this isn't the case.